testhelpers/docker_registry.go

```diff
@@ -64,143 +59,82 @@ func NewDockerRegistry(ops ...RegistryOption) *DockerRegistry {
 	return registry
 }
 
-func (r *DockerRegistry) Start(t *testing.T) {
-	t.Helper()
-
-	r.Host = DockerHostname(t)
-
-	t.Logf("run registry on %s", r.Host)
+func BasicAuth(handler http.Handler, username, password, realm string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		user, pass, ok := r.BasicAuth()
 
-	PullIfMissing(t, DockerCli(t), registryImageName)
+		if !ok || user != username || pass != password {
+			w.Header().Set("WWW-Authenticate", `Basic realm="`+realm+`"`)
+			w.WriteHeader(401)
+			_, _ = w.Write([]byte("Unauthorised.\n"))
+			return
+		}
 
-	registryEnv := []string{
-		"REGISTRY_STORAGE_DELETE_ENABLED=true",
-	}
+		handler.ServeHTTP(w, r)
+	})
+}
 
-	var htpasswdTar io.ReadCloser
-	if r.username != "" {
-		// Create htpasswdTar and configure registry env
-		tempDir, err := ioutil.TempDir("", "test.registry")
-		AssertNil(t, err)
-		defer os.RemoveAll(tempDir)
-
-		htpasswdTar = generateHtpasswd(t, tempDir, r.username, r.password)
-		defer htpasswdTar.Close()
-
-		otherEnvs := []string{
-			"REGISTRY_AUTH=htpasswd",
-			"REGISTRY_AUTH_HTPASSWD_REALM=Registry Realm",
-			"REGISTRY_AUTH_HTPASSWD_PATH=/registry_test_htpasswd",
+func ReadOnly(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !(r.Method == "GET" || r.Method == "HEAD") {
+			w.WriteHeader(405)
+			_, _ = w.Write([]byte("Method Not Allowed.\n"))
+			return
 		}
-		registryEnv = append(registryEnv, otherEnvs...)
-	} else {
-		// make read-only without auth
-		readOnlyEnv := `REGISTRY_STORAGE_MAINTENANCE_READONLY={"enabled":true}`
-		registryEnv = append(registryEnv, readOnlyEnv)
-	}
 
-	var volumeBinds []string
-	var containerUser string
-	if r.volumeName != "" {
-		// try to create volumes that may exist
-		_, err := DockerCli(t).VolumeCreate(context.Background(), volumetypes.VolumeCreateBody{Name: r.volumeName})
-		if err != nil {
-			// fail if err is not from existing volume
-			if !errdefs.IsConflict(err) {
-				AssertNil(t, err)
-			}
-		}
+		handler.ServeHTTP(w, r)
+	})
+}
 
-		info, err := DockerCli(t).Info(context.Background())
-		AssertNil(t, err)
+func (r *DockerRegistry) Start(t *testing.T) {
+	t.Helper()
 
-		storageBindPath := "/registry-storage"
-		if info.OSType == "windows" {
-			containerUser = "ContainerAdministrator" //required for volume permissions
-			storageBindPath = "c:/registry-storage"
-		}
+	r.Host = DockerHostname(t)
 
-		volumeBinds = append(volumeBinds, fmt.Sprintf("%s:%s", r.volumeName, storageBindPath))
+	// create registry handler, if not re-using a shared one
+	if r.regHandler == nil {
+		// change to os.Stderr for verbose output
```

Comment:
Is this a thing? Do we give this knob to the consumer?

Reply:
`regHandler` would only have been set by `WithSharedHandler`: https://github.com/buildpacks/imgutil/pull/110/files/f4c144df5e3abddb910e47e17b95b64bf82c7af4#diff-66c9268014342d51b1c5446732d8eb82958f0f8c1c04e7199033662cabd8a35cR37